### PR TITLE
neutron: enable designate integration

### PR DESCRIPTION
**Pull request by sjamgade**
the config enabled by this flag lets neutron create dns records when
floating ip are created and assigned.
The config was previously added but was not enabled as the barclamp was
still under development 

This config is not really needed for designate or neutron to work, but they help enable certain features
this is where the flag is used https://github.com/crowbar/crowbar-openstack/blob/a671c048ec071a13170d055b4dfc5a6906596cd0/chef/cookbooks/neutron/templates/default/neutron.conf.erb#L36-L48

### CODE REVIEW

**Comment by KeithMnemonic** [comment]
But is the neutron barclamp creating the ext-net ? If so you need to see the dns_name on that network. Is that also being done? And are you using the ml2 driver? that is another step needed.

**Comment by sjamgade** [comment]
Thanks for input, I was not aware of these steps. let me wip this

**Comment by dirkmueller** [comment]
there is a pending XXX for this for floating (this is the name of ext-net)

**Comment by jgrassler** [comment]
What is a "pending XXX"?

File: chef/cookbooks/neutron/recipes/server.rb [hunk 1]
```diff
@@ -445,3 +445,27 @@
     only_if { ha_enabled }
   end
 end
+
+floating_dns_domain = node[:neutron][:floating_dns_domain]
```
Comment (dirkmueller) [inline chef/cookbooks/neutron/recipes/server.rb]: put all of this in post_install_conf after the floating etc is being created? otherwise execution order might be wrong.

**Comment by dirkmueller** [comment]
@jgrassler https://github.com/crowbar/crowbar-openstack/pull/2033/commits/7ca9b937c9a49f0b7de956e72fd6583305e10ecb

File: chef/cookbooks/neutron/recipes/post_install_conf.rb [hunk 2]
```diff
@@ -236,3 +236,25 @@ def mask_to_bits(mask)
   notifies :run, "execute[Neutron network configuration]", :delayed
 end
 
+floating_dns_domain = node[:neutron][:floating_dns_domain]
+env = {}
+env["OS_USERNAME"] = keystone_settings["service_user"]
```
Comment (dirkmueller) [inline chef/cookbooks/neutron/recipes/post_install_conf.rb]: this is already on line 73ff, can we reuse that? I don't see the difference.

File: chef/cookbooks/neutron/recipes/post_install_conf.rb [hunk 3]
```diff
@@ -236,3 +236,25 @@ def mask_to_bits(mask)
...
+env["OS_PASSWORD"] = keystone_settings["service_password"]
+env["OS_PROJECT_NAME"] = keystone_settings["service_tenant"]
+env["OS_AUTH_URL"] = keystone_settings["internal_auth_url"]
+env["OS_REGION_NAME"] = keystone_settings["endpoint_region"]
+env["OS_IDENTITY_API_VERSION"] = keystone_settings["api_version"]
+env["OS_INTERFACE"] = "internal"
+env["OS_USER_DOMAIN_NAME"] = "Default"
+env["OS_PROJECT_DOMAIN_NAME"] = "Default"
+
+if CrowbarOpenStackHelper.config_for_role_exists?("designate")
```
Comment (dirkmueller) [inline chef/cookbooks/neutron/recipes/post_install_conf.rb]: why do we do this only when designate exists?

File: chef/cookbooks/neutron/recipes/post_install_conf.rb [hunk 4]
```diff
@@ -236,3 +236,25 @@ def mask_to_bits(mask)
...
+  bash "Update dns for external network" do
```
Comment (dirkmueller) [inline chef/cookbooks/neutron/recipes/post_install_conf.rb]: please add this as a phase 2 tasks in the "execute block" on line 215ff after floating etc has been created.

File: chef/cookbooks/neutron/recipes/post_install_conf.rb [hunk 5]
```diff
@@ -236,3 +236,25 @@ def mask_to_bits(mask)
...
+    user "root"
+    env env
+    code "
+      openstack network show floating -f value -c dns_domain |\
```
Comment (dirkmueller) [inline chef/cookbooks/neutron/recipes/post_install_conf.rb]: this could be a only_if statement.

File: chef/cookbooks/neutron/recipes/common_config.rb [hunk 6]
```diff
@@ -119,6 +119,18 @@
   infoblox_settings = neutron[:neutron][:infoblox]
 end
 
+designate_public_uri = nil
+designate_enabled = false
+if CrowbarOpenStackHelper.config_for_role_exists?("designate")
+  designate_enabled = true
+  node_designate = node_search_with_cache("roles:designate-server").first[:designate]
+  public_host = CrowbarHelper.get_host_for_public_url(node,
```
Comment (dirkmueller) [inline chef/cookbooks/neutron/recipes/common_config.rb]: this should be node_designate.

**Comment by jgrassler** [comment]
> @jgrassler [7ca9b93](https://github.com/crowbar/crowbar-openstack/commit/7ca9b937c9a49f0b7de956e72fd6583305e10ecb)

Interesting. i just looked at what Sumit's got and that config section lacked domain information. Upon further investigation we stumbled upon this little gem:

https://github.com/openstack/neutron/blob/stable/rocky/neutron/services/externaldns/drivers/designate/driver.py#L54

tl;dr: the Designate integration in Neutron does not support Keystone V2. We'll try to come up with a patch.

**Comment by toabctl** [comment]
> tl;dr: the Designate integration in Neutron does not support Keystone V2. We'll try to come up with a patch.

I think you can just set the domain info and set `auth_type = password` (then you land here: https://github.com/openstack/neutron/blob/stable/rocky/neutron/services/externaldns/drivers/designate/driver.py#L50). iirc using `password` and having the needed info will automatically select the correct version. But you could also set `auth_type = v3password` to be sure to use v3. @cmurphy might know more about this...

**Comment by cmurphy** [comment]
Yes, just use auth_type=password and set user_domain_name and project_domain_name

File: chef/cookbooks/neutron/recipes/post_install_conf.rb [hunk 7]
```diff
@@ -211,6 +211,16 @@ def mask_to_bits(mask)
   action :nothing
 end
 
+doamin_floating = node[:neutron][:floating_dns_domain]
+
+execute "update_dns_domain_for_floating_network" do
+  command "#{env} neutron net-update floating --dns-domain #{doamin_floating}"
+  not_if "#{openstack_cmd} network show floating -f value -c dns_domain | grep -q #{doamin_floating}"
```
[the remark this answers is not in the document]
Comment (sjamgade) [reply chef/cookbooks/neutron/recipes/post_install_conf.rb]: 1 char more should be fine

File: chef/cookbooks/neutron/recipes/common_config.rb [hunk 8]
```diff
@@ -119,6 +119,19 @@
   infoblox_settings = neutron[:neutron][:infoblox]
 end
 
+designate_public_uri = nil
+designate_enabled = false
+if CrowbarOpenStackHelper.config_for_role_exists?("designate")
+  designate_enabled = true
+  designate_server = node_search_with_cache("roles:designate-server").first
```
Comment (dirkmueller) [inline chef/cookbooks/neutron/recipes/common_config.rb]: as you're doing a node_search_with_cache anyway, you can skip the "config_for_role" part completely - you're using anyway the node attributes. then you can do this:

```
  designate_public_uri = nil
  designate_server = node_search_with_cache("roles:designate-server").first
  unless designate_server.nil?
    ...

   designate_enabled: not designate_server.nil?

```

Comment (sjamgade) [reply chef/cookbooks/neutron/recipes/common_config.rb]: Done
@dirkmueller could you please re-approve

**Comment by sjamgade** [comment]
This should be ready for reviews !

File: chef/cookbooks/neutron/recipes/post_install_conf.rb [hunk 9]
```diff
@@ -211,6 +211,25 @@ def mask_to_bits(mask)
   action :nothing
 end
 
+doamin_floating = node[:neutron][:floating_dns_domain]
```
Comment (jgrassler) [inline chef/cookbooks/neutron/recipes/post_install_conf.rb]: Typo: "doamin" and it's pretty consistent throughout the section...any reason for that?

File: chef/cookbooks/neutron/recipes/post_install_conf.rb [hunk 10]
```diff
@@ -226,6 +245,8 @@ def mask_to_bits(mask)
   notifies :run, "execute[set_router_gateway]", :delayed
   notifies :run, "execute[add_fixed_network_to_router]", :delayed
   notifies :run, "execute[add_ironic_network_to_router]", :delayed
+  notifies :run, "execute[update_dns_domain_for_fixed_network]", :delayed
```
Comment (dirkmueller) [inline chef/cookbooks/neutron/recipes/post_install_conf.rb]: I believe this should only be done when designate integration is enabled?

File: chef/cookbooks/neutron/recipes/common_config.rb [hunk 11]
```diff
@@ -119,6 +119,16 @@
...
 end
 
+designate_server = node_search_with_cache("roles:designate-server").first
+unless designate_server.nil?
+  node_designate = designate_server[:designate]
+  public_host = CrowbarHelper.get_host_for_public_url(designate_server,
+                                                    node_designate[:api][:protocol] == "https",
+                                                    node_designate[:ha][:enabled])
+  api_protocol = node_designate[:api][:protocol]
+  designate_public_uri = "#{api_protocol}://#{public_host}:#{node_designate[:api][:bind_port]}/v2"
```
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/common_config.rb]: `designate_public_uri` is not defined outside of this block; I'd like to have an explicit `designate_public_uri = nil` for the other case

File: chef/data_bags/crowbar/template-neutron.json [hunk 12]
```diff
@@ -10,8 +10,8 @@
       "max_header_line": 16384,
       "debug": false,
       "create_default_networks": true,
-      "dns_domain": "openstack.local",
-      "floating_dns_domain": "floating.openstack.local",
+      "dns_domain": "openstack.local.",
```
Comment (vuntz) [inline chef/data_bags/crowbar/template-neutron.json]: Should we have a migration to add this `.` at the end of the attribute, for upgrades?

File: chef/cookbooks/neutron/recipes/post_install_conf.rb (hunk 10 above)
Comment (sjamgade) [reply chef/cookbooks/neutron/recipes/post_install_conf.rb]: not necessary, it can be done otherwise too, putting it in there just makes it explicit

File: chef/cookbooks/neutron/recipes/common_config.rb (hunk 11 above)
Comment (sjamgade) [reply chef/cookbooks/neutron/recipes/common_config.rb]: last time I checked that was implicit,

File: chef/data_bags/crowbar/template-neutron.json (hunk 12 above)
Comment (KeithMnemonic) [reply chef/data_bags/crowbar/template-neutron.json]: You also need to make sure the "." is in the neutron.conf override or wherever it is configured. https://docs.openstack.org/mitaka/networking-guide/config-dns-int.html

Comment (dirkmueller) [reply chef/data_bags/crowbar/template-neutron.json]: or just fix it up as part of the chef recipe.. I am not sure we want to dns propagate a domain called "openstack.local" anyway (or anything else that is *local)

File: chef/cookbooks/neutron/recipes/common_config.rb (hunk 11 above)
Comment (sjamgade) [reply chef/cookbooks/neutron/recipes/common_config.rb]: done

File: chef/data_bags/crowbar/template-neutron.json (hunk 12 above)
Comment (sjamgade) [reply chef/data_bags/crowbar/template-neutron.json]: added one for both the domains

File: chef/cookbooks/neutron/recipes/common_config.rb [hunk 13]
```diff
@@ -119,6 +119,17 @@
   infoblox_settings = neutron[:neutron][:infoblox]
 end
 
+designate_public_uri = nil
+designate_server = node_search_with_cache("roles:designate-server").first
```
Comment (cmurphy) [inline chef/cookbooks/neutron/recipes/common_config.rb]: What if there are multiple nodes, as in an HA case?

Comment (jgrassler) [reply chef/cookbooks/neutron/recipes/common_config.rb]: That doesn't pose a problem: we just need a node object for any node with the designate role. A little further down we invoke `CrowbarHelper.get_host_for_public_url()` on that node object and that function will Do The Right Thing on an HA cloud and yield the HA cluster's public host (I've checked in QA's HA enabled test environment).

Comment (cmurphy) [reply chef/cookbooks/neutron/recipes/common_config.rb]: Okay, thanks!

File: chef/cookbooks/neutron/recipes/common_config.rb [hunk 14]
```diff
@@ -132,8 +142,8 @@
       sql_max_pool_overflow: neutron[:neutron][:sql][:max_pool_overflow],
       sql_pool_timeout: neutron[:neutron][:sql][:pool_timeout],
       debug: neutron[:neutron][:debug],
-      # TODO
-      designate_enabled: false,
+      designate_enabled: (not designate_server.nil?),
```
[the remark this answers is not in the document]
Comment (cmurphy) [reply chef/cookbooks/neutron/recipes/common_config.rb]: If you end up needing to alter this PR again you may as well fix this hound complaint
